### ISSUE TITLE
[Research] Find Python 3 Compatible Replacement for MongoKit

### ISSUE THREAD

**Issue by mcdonnnj** [cisagov/cyhy-core#52]
This project currently relies on the [MongoKit](https://github.com/namlook/mongokit) package to provide ODM functionality. That project was never made Python 3 compatible, nor was it updated for later versions of `PyMongo`. As it is now archived, that is likely to never happen. A suitable replacement needs to be picked to replace it, and the code in this repository updated, so that Python 3 migration can progress.

A compare and contrast needs to be done between options like [MongoKat](https://github.com/pricingassistant/mongokat) and the projects listed in the MongoDB [tools page](https://api.mongodb.com/python/current/tools.html#orm-like-layers). It would be ideal if it operated in a fashion similar to `MongoKit` while still retaining similar features to ease the transition.

**Comment by mcdonnnj** [comment]
# Rough Comparisons #

## MongoKit (what we use now) ##

* Readily exposes the underlying PyMongo object.
* Does not add fields of its own to documents.
* Override save() method to apply changes when saving a document.

## PyMODM ##

* Does not freely expose underlying PyMongo object, but does support raw queries.
* Adds a _cls field for inheritence in the same collection, but might to be disable-able through the `final` metadata attribute.
* Will need to set the `collection_name` metadata attribute of each Document model to set the collection name to use.
* Implement abstract method clean() to apply changes when saving a document.
* Will need to make a custom QuerySet.
* Queries should be restructured to fit the PyMODM design.
* Use of exceptions instead of returning None for some search methods.

## uMongo ##

* Readily exposes the underlying PyMongo object.
* Adds a _cls field to track model inheritence, may be configurable with the `is_child` metadata attribute.
* Will need to set the `collection_name` metadata attribute of each Document model to set the collection name to use.
* Changes when saving may be doable through pre_insert()/pre_update() methods?

## MongoEngine ##

* Does not freely expose underlying PyMongo object, but does support raw queries.
* Adds a _cls field for inheritence in the same collection, but appears to be disableable through the `allow_inheritance` metadata attribute.
* Will need to set the `collection` metadata attribute of each Document model to set the collection name to use.
* Implement clean() method to apply changes when saving a document.
* Will need to make a custom QuerySet.
* Queries should be restructured to fit the MongoEngine design.
* Possible migration issues if a field is added to a document's schema.
* Use of exceptions instead of returning None for some search methods.

**Comment by mcdonnnj** [comment]
After discussion, the similarities of PyMODM and MongoEngine left the main argument down to support. The pause of PyMODM maintenance per [this forum post](https://developer.mongodb.com/community/forums/t/updates-on-pymodm/9363) was a major reason to side with MongoEngine.

**Comment by mcdonnnj** [comment]
We moved on to discussing whether the transition to Python 3 should happen simultaneously with the transition to MongoEngine. We decided to switch to MongoEngine first, and then work on the transition to Python 3 after for the following reasons:
* The last release of MongoEngine that supported Python 2 was [v0.19.1](https://github.com/MongoEngine/mongoengine/releases/tag/v0.19.1) in 2020-01-04, so it is relatively recent.
* It is easier to add a Python 3 migration mid-stream if necessary than to decouple a Python 3 migration mid-stream.
* Given how interconnected the code in this repository is, and the number of other projects that rely on it, it is a saner approach to transition to Python 3 after a MongoEngine transition.
